File: src/Polyball/polyball.cpp
//
// Created by dydxh on 1/8/19.
//

#include "polyball.h"

Polyball::Polyball(Shader shader) {
    Speed = glm::vec3(0.0f, 0.0f, 0.0f);
    SpeedLimit = glm::vec3(0.0f);
    Acceleration = glm::vec3(2.0f, 0.0f, 2.0f);
    Resistance = glm::vec3(1.0f, 3.0f, 1.0f);
    Position = glm::vec3(0.0f, 2.0f, 5.0f);
    Radius = 0.5f * 0.3f;
    Mov.Front = glm::vec3(1.0f, 0.0f, 0.0f);
    Mov.Right = glm::vec3(0.0f, 0.0f, 1.0f);
    Mov.Up    = glm::vec3(0.0f, 1.0f, 0.0f);
    Cam.Front = ResourceManager::camera.Front;
    Cam.Right = ResourceManager::camera.Right;
    Cam.Up    = ResourceManager::camera.Up;

    collision = (Collision){false, glm::vec3(0.0f), glm::vec3(0.0f)};
    rotate_state = glm::quat(1.0f, 0.0f, 0.0f, 0.0f);

    BallShader = shader;
    Initialize();
}

Polyball::~Polyball() {}

void Polyball::Initialize() {
    modelptr = ResourceManager::GetModel("polyball");
}

glm::vec3 Polyball::GenCameraPosition() {
    const GLfloat distance = 3.0f;
    static GLfloat cosvalue = glm::cos(PI / 12.0);
    static GLfloat sinvalue = glm::sin(PI / 12.0);
    glm::vec3 ret = Position;
    ret += -distance * cosvalue * Cam.Front;
    ret += distance * sinvalue * Cam.Up;
    return ret;
}

void Polyball::Render(glm::mat4 view, glm::mat4 PVMatrix, glm::mat4 lightSpaceMatrix, Texture2D BlurShadow) {
    glm::mat4 model = glm::mat4(1.0f);
    model = glm::translate(model, Position);
    model = glm::scale(model, glm::vec3(0.3f));

    BallShader.use();
    BallShader.setMat4("view", view);
    BallShader.setMat4("PVMatrix", PVMatrix);
    BallShader.setMat4("lightSpaceMatrix", lightSpaceMatrix);
    BallShader.setMat4("model", model);
    BallShader.setVec4("RotQuat", glm::vec4(rotate_state.w, rotate_state.x, rotate_state.y, rotate_state.z));
    BallShader.setLight(ResourceManager::camera.Position);

    glActiveTexture(GL_TEXTURE0);
    BlurShadow.Bind();

    modelptr->Draw(BallShader);

//    Tools::PrintVec3(Position);
//    Tools::PrintVec3(Speed);
}
glm::vec3 Polyball::CollisionObject(Scene *scene) {
    glm::vec3 delta_pos = glm::vec3(0.0f);
    for(int typenum = 0; typenum < 3; typenum++){
        for(int i = 0; i < scene->Treeplace[typenum].size(); i++){
            if(typenum <= 1){
                Trunk tk = scene->Treeplace[typenum][i].trunk;
                GLfloat seglen = tk.height;
                glm::vec3 seg = glm::vec3(0.0f, 1.0f, 0.0);
                glm::vec3 vect = Position - tk.bottom;
                GLfloat len = glm::dot(seg, vect);
                glm::vec3 Q = len * seg + tk.bottom;
                GLfloat mindist;
                if(len < 0) mindist = glm::length(vect);
                else if(len > seglen) mindist = glm::length(Position - (tk.bottom + glm::vec3(0.0f, tk.height, 0.0f)));
                else mindist = glm::length(Position - Q);
                if(mindist < Radius + tk.Radius){
                    collision.exist = true;
                    glm::vec3 normal = glm::normalize(glm::vec3(Position.x - tk.bottom.x, 0.0f, Position.z - tk.bottom.z));
                    collision.Normal += normal;
                    delta_pos += (Radius + tk.Radius - mindist) * normal;
                }
            }
            else{
                BoundBox box = scene->Treeplace[typenum][i].box;
                glm::vec3 cp = box.Point;// - glm::vec3(box.Left, box.Down, box.Front) / 2.0f;
                if(Tools::distance(Position, cp) >= 1.1) continue;
                GLfloat test = Tools::distance(Position, cp);

                glm::vec3 relpos = Position - cp;
                GLfloat disx = glm::abs(relpos.x) - box.Left / 2.0f;
                GLfloat disy = glm::abs(relpos.y) - box.Down / 2.0f;
                GLfloat disz = glm::abs(relpos.z) - box.Front / 2.0f;
                GLfloat dist = glm::sqrt(glm::pow(glm::max(0.0f, disx), 2) +
                                         glm::pow(glm::max(0.0f, disy), 2) +
                                         glm::pow(glm::max(0.0f, disz), 2));
                if(dist < Radius){
                    collision.exist = true;
                    glm::vec3 normal;
                    if(disx >= disy && disx >= disz){
                        if(relpos.x > 0) normal = glm::vec3(1.0f, 0.0f, 0.0f);
                        else normal = glm::vec3(-1.0f, 0.0f, 0.0f);
                    }
                    else if(disy >= disx && disy >= disz){
                        if(relpos.y > 0) normal = glm::vec3(0.0f, 1.0f, 0.0f);
                        else normal = glm::vec3(0.0f, -1.0f, 0.0f);
                    }
                    else{
                        if(relpos.z > 0) normal = glm::vec3(0.0f, 0.0f, 1.0f);
                        else normal = glm::vec3(0.0f, 0.0f, -1.0f);
                    }
                    collision.Normal += normal;
                    delta_pos += normal * (Radius - dist);
                }
            }
        }
    }
    return delta_pos;
}
void Polyball::CollisionCheck(Scene* scene) {
    collision = (Collision){false, glm::vec3(0.0f), glm::vec3(0.0f)};
//    if(Position.y <= 0.0f){
//        Position.y = 0.0f;
//        collision = (Collision){true, glm::vec3(0.0f), glm::vec3(0.0f, 1.0f, 0.0f)};
//    }
    GLint cx, cy, mx, my;
    GetChunkMeshID(scene, cx, cy, mx, my);
    dcx = cx, dcy = cy, dmx = mx, dmy = my;
    for(int i = -1; i <= 2; i++)
        for(int j = -1; j <= 2; j++){
            int meshx = mx + i, meshy = my + j;
            int chunkx, chunky;
            if(meshx < 0) chunkx = cx - 1, meshx += MESH_SIZE;
            else if(meshx >= MESH_SIZE) chunkx = cx + 1, meshx -= MESH_SIZE;
            else chunkx = cx;
            if(meshy < 0) chunky = cy - 1, meshy += MESH_SIZE;
            else if(meshy >= MESH_SIZE) chunky = cy + 1, meshy -= MESH_SIZE;
            else chunky = cy;
            Maycol[i + 1][j + 1] = scene->chunk_offset[chunkx][chunky] + scene->mesh_offset[meshx][meshy];
            Maycol[i + 1][j + 1].y = scene->chunk[chunkx][chunky]->height[meshx][meshy];
            if(Maycol[i + 1][j + 1].y < 0.1f) Maycol[i + 1][j + 1].y = -0.5f;
        }
    glm::vec3 delta_pos = glm::vec3(0.0f);
    for(int i = 0; i < 3; i++){
        for(int j = 0; j < 3; j++){
            glm::vec3 a, b, c;
            GLfloat dist;
            a = Maycol[i][j];
            b = Maycol[i + 1][j];
            c = Maycol[i + 1][j + 1];
            dist = Tools::distance(Position, a, b, c);
            if(dist <= Radius){
                collision.exist = true;
                glm::vec3 normal = glm::normalize(glm::cross(c - a, b - a));
                collision.Normal += normal;
                delta_pos += (Radius - dist) * normal;
            }
            b = c;
            c = Maycol[i][j + 1];
            dist = Tools::distance(Position, a, b, c);
            if(dist <= Radius){
                collision.exist = true;
                glm::vec3 normal = glm::normalize(glm::cross(c - a, b - a));
                collision.Normal += normal;
                delta_pos += (Radius - dist) * normal;
            }
        }

    }
    delta_pos += CollisionObject(scene);
    if(Position.y < 0.1){
        collision.exist = true;
        collision.Normal += glm::vec3(0.0f, 1.0f, 0.0f);
        delta_pos += (0.1f - Position.y) * glm::vec3(0.0f, 1.0f, 0.0f);
    }
    Position += delta_pos;
    if(collision.exist)
        collision.Normal = glm::normalize(collision.Normal);
}
void Polyball::GetChunkMeshID(Scene* scene, GLint &cx, GLint &cz, GLint &mx, GLint &mz){
    glm::vec3 position = Position -
                         scene->chunk[CHUNK_RADIUS][CHUNK_RADIUS]->submesh[MESH_RADIUS][MESH_RADIUS]->get_Position() -
                         glm::vec3(MESH_LENGTH / 2.0f, 0.0f, MESH_LENGTH / 2.0f);

    if(position.x > 0)
        cx = static_cast<GLint>(position.x / CHUNK_LENGTH + 0.5f) + CHUNK_RADIUS;
    else
        cx = static_cast<GLint>(position.x / CHUNK_LENGTH - 0.5f) + CHUNK_RADIUS;
    if(position.z > 0)
        cz = static_cast<GLint>(position.z / CHUNK_LENGTH + 0.5f) + CHUNK_RADIUS;
    else
        cz = static_cast<GLint>(position.z / CHUNK_LENGTH - 0.5f) + CHUNK_RADIUS;
    position.x -= (cx - CHUNK_RADIUS) * CHUNK_LENGTH;
    position.z -= (cz - CHUNK_RADIUS) * CHUNK_LENGTH;
    if(position.x > 0)
        mx = static_cast<GLint>(position.x / MESH_LENGTH + 0.5f) + MESH_RADIUS;
    else
        mx = static_cast<GLint>(position.x / MESH_LENGTH - 0.5f) + MESH_RADIUS;
    if(position.z > 0)
        mz = static_cast<GLint>(position.z / MESH_LENGTH + 0.5f) + MESH_RADIUS;
    else
        mz = static_cast<GLint>(position.z / MESH_LENGTH - 0.5f) + MESH_RADIUS;

}

//UpdatePosition => Process Speed
void Polyball::UpdatePosition(float deltaTime, Scene* scene) {
    Position += Speed * deltaTime;
    CollisionCheck(scene);

    GLfloat angle = wspeed * deltaTime / 2.0f;
    glm::quat rot = glm::quat(glm::cos(angle),
                    glm::sin(angle) * rotate_axis.x,
                    glm::sin(angle) * rotate_axis.y,
                    glm::sin(angle) * rotate_axis.z);
    rotate_state = rot * rotate_state;
}

//UpdateSpeed => Process Acceleration
void Polyball::UpdateSpeed(float deltaTime) {
    if(collision.exist){
        GLfloat velocity = -glm::dot(collision.Normal, Speed);
        Speed += velocity * collision.Normal;
        if(velocity > 2.0f)
            Speed += 0.1f * velocity * collision.Normal;
    }
    UpdateMovVec();

    glm::vec3 delta_v = glm::vec3(0.0f);
    bool noKeyboard = true;
    if(collision.exist){
        if(ResourceManager::Keys[GLFW_KEY_W]){
            GLfloat cosvalue = glm::dot(Cam.Front, Mov.Up);
            delta_v += (Cam.Front - cosvalue * Mov.Up) * Acceleration.x * deltaTime;
            noKeyboard = false;
        }
        if(ResourceManager::Keys[GLFW_KEY_S]){
            GLfloat cosvalue = glm::dot(Cam.Front, Mov.Up);
            delta_v -= (Cam.Front - cosvalue * Mov.Up) * Acceleration.x * deltaTime;
            noKeyboard = false;
        }
        if(ResourceManager::Keys[GLFW_KEY_D]){
            GLfloat cosvalue = glm::dot(Cam.Right, Mov.Up);
            delta_v += (Cam.Right - cosvalue * Mov.Up) * Acceleration.z * deltaTime;
            noKeyboard = false;
        }
        if(ResourceManager::Keys[GLFW_KEY_A]){
            GLfloat cosvalue = glm::dot(Cam.Right, Mov.Up);
            delta_v -= (Cam.Right - cosvalue * Mov.Up) * Acceleration.z * deltaTime;
            noKeyboard = false;
        }
        if(ResourceManager::Keys[GLFW_KEY_SPACE]){
            delta_v += glm::vec3(0.0f, 3.0f, 0.0f);
        }
        //Sliding Friction
        glm::vec3 movXZ = Speed;
        GLfloat movSpeed = glm::length(movXZ);
        if(movSpeed > eps) {
            if (Resistance.x * deltaTime > movSpeed && noKeyboard)
                delta_v -= glm::normalize(movXZ) * movSpeed;
            else
                delta_v -= glm::normalize(movXZ) * Resistance.x * deltaTime;
        }
    }
    //Gravity
    delta_v.y -= Resistance.y * deltaTime;
    Speed += delta_v;

    UpdateMovVec();

    if(glm::length(glm::vec2(Speed.x, Speed.z)) < 0.001) wspeed = 0.0f;
    else wspeed = glm::length(Speed) / Radius;
    rotate_axis = -Mov.Right;
}
void Polyball::UpdateMovVec(){
    if(glm::abs(Speed.x) < eps && glm::abs(Speed.z) < eps){
        Mov.Front = glm::vec3(1.0f, 0.0f, 0.0f);
        Mov.Right = glm::vec3(0.0f, 0.0f, 1.0f);
        Mov.Up    = glm::vec3(0.0f, 1.0f, 0.0f);
    }
    else{
        Mov.Front = glm::normalize(Speed);
        Mov.Right = glm::normalize(glm::cross(Mov.Front, glm::vec3(0.0f, 1.0f, 0.0f)));
        Mov.Up    = glm::normalize(glm::cross(Mov.Right, Mov.Front));
    }
}

File: iOS/Polyworld/Model/Polyball.swift
//
//  Polyball.swift
//  Polyworld
//
//  Created by 沈心逸 on 2019/1/11.
//  Copyright © 2019年 Sxy. All rights reserved.
//

import Foundation
import MetalKit
import simd

struct Collision{
    var exist: Bool
    var Point: float3
    var Normal: float3
}

struct Coord{
    var Front: float3
    var Right: float3
    var Up: float3
}

class Polyball: NSObject {
    let device: MTLDevice
    let commandQueue: MTLCommandQueue
    let renderPipeline: MTLRenderPipelineState
    let depthStencilState: MTLDepthStencilState
    let vertexDescriptor: MDLVertexDescriptor
    
    let textureLoader: MTKTextureLoader
    let defaultTexture: MTLTexture
    let defaultNormalMap: MTLTexture
    //let irradianceCubeMap: MTLTexture
    
    var nodes = [Node]()
    
    var modelMatrix = matrix_identity_float4x4
    var viewMatrix = matrix_identity_float4x4
    var projectionMatrix = matrix_identity_float4x4
    var cameraWorldPosition = float3(0, 0, 0)
    var cameraFront =  float3(0, 0, -1)
    var lightWorldDirection = float3(0, 1, 0)
    var lightWorldPosition = float3(0, 5, 0)
    var time: Float = 0
    
    var speed: float3 = float3(0.0, 0.0, 0.0)
    var speedLimit: float3 = float3(0.0, 0.0, 0.0)
    var Acceleration: float3 = float3(2.0, 0.0, 2.0)
    var Resistance: float3 = float3(1.0, 3.0, 1.0)
    var Position: float3 = float3(0.0, 0.01, 4.0)
    var Radius: Float = 0.5 * 0.3
    var Mov: Coord = Coord(Front: float3(1.0, 0.0, 0.0), Right: float3(0.0, 0.0, 1.0), Up: float3(0.0, 1.0, 0.0))
    var Cam: Coord = Coord(Front: ResourceManager.camera.front, Right: ResourceManager.camera.right, Up: ResourceManager.camera.up)
    var collision: Collision = Collision(exist: false, Point: float3(), Normal: float3())
    
    var wspeed: Float!
    var rotate_axis: float3!
    var rotate_state: simd_quatf = simd_quatf(real: 1.0, imag: float3())
    
    var Maycol = [[float3]](repeating: [float3](repeating: float3(0,0,0), count: 5), count: 5)
    
    var dcx,dcy,dmx,dmy: Int!
    
    init(device: MTLDevice, modelMatrix: float4x4, forResourse name: String, withExtension ext: String) {
        self.device = device
        self.modelMatrix = modelMatrix
        commandQueue = device.makeCommandQueue()!
        vertexDescriptor = Polyball.buildVertexDescriptor(device: device)
        renderPipeline = Polyball.buildPipeline(device: device, vertexDescriptor: vertexDescriptor)
        depthStencilState = Polyball.buildDepthStencilState(device: device)
        textureLoader = MTKTextureLoader(device: device)
        (defaultTexture, defaultNormalMap) = Polyball.buildDefaultTextures(device: device)
        //irradianceCubeMap = Renderer.buildEnvironmentTexture("garage_pmrem.ktx", device: device)
        super.init()
        
        guard let modelURL = Bundle.main.url(forResource: name, withExtension: ext) else {
            fatalError("Could not find model file in app bundle")
        }
        buildScene(url: modelURL, device: device, vertexDescriptor: vertexDescriptor)
    }
    
    static func buildVertexDescriptor(device: MTLDevice) -> MDLVertexDescriptor {
        let vertexDescriptor = MDLVertexDescriptor()
        vertexDescriptor.attributes[0] = MDLVertexAttribute(name: MDLVertexAttributePosition,
                                                            format: .float3,
                                                            offset: 0,
                                                            bufferIndex: VertexBufferIndex.attributes.rawValue)
        vertexDescriptor.attributes[1] = MDLVertexAttribute(name: MDLVertexAttributeNormal,
                                                            format: .float3,
                                                            offset: MemoryLayout<Float>.size * 3,
                                                            bufferIndex: VertexBufferIndex.attributes.rawValue)
        vertexDescriptor.attributes[2] = MDLVertexAttribute(name: MDLVertexAttributeTangent,
                                                            format: .float3,
                                                            offset: MemoryLayout<Float>.size * 6,
                                                            bufferIndex: VertexBufferIndex.attributes.rawValue)
        vertexDescriptor.attributes[3] = MDLVertexAttribute(name: MDLVertexAttributeTextureCoordinate,
                                                            format: .float2,
                                                            offset: MemoryLayout<Float>.size * 9,
                                                            bufferIndex: VertexBufferIndex.attributes.rawValue)
        vertexDescriptor.layouts[VertexBufferIndex.attributes.rawValue] = MDLVertexBufferLayout(stride: MemoryLayout<Float>.size * 11)
        return vertexDescriptor
    }
    
    static func buildPipeline(device: MTLDevice, vertexDescriptor: MDLVertexDescriptor) -> MTLRenderPipelineState {
        guard let library = device.makeDefaultLibrary() else {
            fatalError("Could not load default library from main bundle")
        }
        
        let vertexFunction = library.makeFunction(name: "polyballVertex")
        let fragmentFunction = library.makeFunction(name: "polyballFragment")
        
        let pipelineDescriptor = MTLRenderPipelineDescriptor()
        pipelineDescriptor.vertexFunction = vertexFunction
        pipelineDescriptor.fragmentFunction = fragmentFunction
        
        pipelineDescriptor.colorAttachments[0].pixelFormat = .bgra8Unorm
        pipelineDescriptor.depthAttachmentPixelFormat = .depth32Float
        
        let mtlVertexDescriptor = MTKMetalVertexDescriptorFromModelIO(vertexDescriptor)
        pipelineDescriptor.vertexDescriptor = mtlVertexDescriptor
        
        do {
            return try device.makeRenderPipelineState(descriptor: pipelineDescriptor)
        } catch {
            fatalError("Could not create render pipeline state object: \(error)")
        }
    }
    
    static func buildDepthStencilState(device: MTLDevice) -> MTLDepthStencilState {
        let depthStencilDescriptor = MTLDepthStencilDescriptor()
        depthStencilDescriptor.depthCompareFunction = .less
        depthStencilDescriptor.isDepthWriteEnabled = true
        return device.makeDepthStencilState(descriptor: depthStencilDescriptor)!
    }
    
    static func buildDefaultTextures(device: MTLDevice) -> (MTLTexture, MTLTexture) {
        let bounds = MTLRegionMake2D(0, 0, 1, 1)
        let descriptor = MTLTextureDescriptor.texture2DDescriptor(pixelFormat: .rgba8Unorm,
                                                                  width: bounds.size.width,
                                                                  height: bounds.size.height,
                                                                  mipmapped: false)
        descriptor.usage = .shaderRead
        let defaultTexture = device.makeTexture(descriptor: descriptor)!
        let defaultColor: [UInt8] = [ 0, 0, 0, 255 ]
        defaultTexture.replace(region: bounds, mipmapLevel: 0, withBytes: defaultColor, bytesPerRow: 4)
        let defaultNormalMap = device.makeTexture(descriptor: descriptor)!
        let defaultNormal: [UInt8] = [ 127, 127, 255, 255 ]
        defaultNormalMap.replace(region: bounds, mipmapLevel: 0, withBytes: defaultNormal, bytesPerRow: 4)
        return (defaultTexture, defaultNormalMap)
    }
    
    static func buildEnvironmentTexture(_ name: String, device:MTLDevice) -> MTLTexture {
        let textureLoader = MTKTextureLoader(device: device)
        let options: [MTKTextureLoader.Option : Any] = [:]
        do {
            let textureURL = Bundle.main.url(forResource: name, withExtension: nil)!
            let texture = try textureLoader.newTexture(URL: textureURL, options: options)
            return texture
        } catch {
            fatalError("Could not load irradiance map from asset catalog: \(error)")
        }
    }
    
    func buildScene(url: URL, device: MTLDevice, vertexDescriptor: MDLVertexDescriptor) {
        let bufferAllocator = MTKMeshBufferAllocator(device: device)
        let asset = MDLAsset(url: url, vertexDescriptor: nil, bufferAllocator: bufferAllocator)
        
        asset.loadTextures()
        
        for sourceMesh in asset.childObjects(of: MDLMesh.self) as! [MDLMesh] {
            sourceMesh.addOrthTanBasis(forTextureCoordinateAttributeNamed: MDLVertexAttributeTextureCoordinate,
                                       normalAttributeNamed: MDLVertexAttributeNormal,
                                       tangentAttributeNamed: MDLVertexAttributeTangent)
            sourceMesh.vertexDescriptor = vertexDescriptor
        }
        
        guard let (sourceMeshes, meshes) = try? MTKMesh.newMeshes(asset: asset, device: device) else {
            fatalError("Could not convert ModelIO meshes to MetalKit meshes")
        }
        
        for (sourceMesh, mesh) in zip(sourceMeshes, meshes) {
            var materials = [Material]()
            for sourceSubmesh in sourceMesh.submeshes as! [MDLSubmesh] {
                let material = Material(material: sourceSubmesh.material, textureLoader: textureLoader)
                materials.append(material)
            }
            let node = Node(mesh: mesh, materials: materials)
            nodes.append(node)
        }
    }
    
    func updateScene(drawable: CAMetalDrawable) {
        //time += 1 / Float(view.preferredFramesPerSecond)
        modelMatrix = float4x4(translationBy: Position) * float4x4(rotate_state)
        projectionMatrix = ResourceManager.projectionMatrix
        viewMatrix = ResourceManager.camera.viewMatrix
        cameraWorldPosition = ResourceManager.camera.position
        lightWorldPosition = PARLIGHT_POSITION
        lightWorldDirection = normalize(PARLIGHT_DIR)
    }
    
    func bindTextures(_ material: Material, _ commandEncoder: MTLRenderCommandEncoder) {
        commandEncoder.setFragmentTexture(material.baseColor ?? defaultTexture, index: TextureIndex.baseColor.rawValue)
        commandEncoder.setFragmentTexture(material.metallic ?? defaultTexture, index: TextureIndex.metallic.rawValue)
        commandEncoder.setFragmentTexture(material.roughness ?? defaultTexture, index: TextureIndex.roughness.rawValue)
        commandEncoder.setFragmentTexture(material.normal ?? defaultNormalMap, index: TextureIndex.normal.rawValue)
        commandEncoder.setFragmentTexture(material.emissive ?? defaultTexture, index: TextureIndex.emissive.rawValue)
    }
    
    func draw(drawable: CAMetalDrawable) {
        updateScene(drawable: drawable)
        
        let white = MTLClearColor(red: 1, green: 1, blue: 1, alpha: 1)
        let renderPassDescriptor = MTLRenderPassDescriptor()
        renderPassDescriptor.colorAttachments[0].texture = drawable.texture
        renderPassDescriptor.colorAttachments[0].loadAction = .load
        renderPassDescriptor.colorAttachments[0].clearColor = white
        renderPassDescriptor.colorAttachments[0].storeAction = .store
        
        renderPassDescriptor.depthAttachment.texture = ResourceManager.depthTexture
        renderPassDescriptor.depthAttachment.loadAction = .load
        renderPassDescriptor.depthAttachment.storeAction = .store
        
        let commandEncoder = ResourceManager.commandBuffer.makeRenderCommandEncoder(descriptor: renderPassDescriptor)!
        commandEncoder.setRenderPipelineState(renderPipeline)
        commandEncoder.setDepthStencilState(depthStencilState)
        
        for node in nodes {
            draw(node, in: commandEncoder)
        }
        
        commandEncoder.endEncoding()
    }
    
    func draw(_ node: Node, in commandEncoder: MTLRenderCommandEncoder) {
        let mesh = node.mesh
        
        var uniforms = Uniforms(modelMatrix: modelMatrix,
                                viewMatrix: viewMatrix,
                                projectionMatrix: projectionMatrix,
                                cameraPosition: cameraWorldPosition,
                                lightDirection: lightWorldDirection,
                                lightPosition: lightWorldPosition)
        commandEncoder.setVertexBytes(&uniforms, length: MemoryLayout<Uniforms>.size, index: VertexBufferIndex.uniforms.rawValue)
        commandEncoder.setFragmentBytes(&uniforms, length: MemoryLayout<Uniforms>.size, index: FragmentBufferIndex.uniforms.rawValue)
        
        for (bufferIndex, vertexBuffer) in mesh.vertexBuffers.enumerated() {
            commandEncoder.setVertexBuffer(vertexBuffer.buffer, offset: vertexBuffer.offset, index: bufferIndex)
        }
        
        for (submeshIndex, submesh) in mesh.submeshes.enumerated() {
            let material = node.materials[submeshIndex]
            bindTextures(material, commandEncoder)
            
            let indexBuffer = submesh.indexBuffer
            commandEncoder.drawIndexedPrimitives(type: submesh.primitiveType,
                                                 indexCount: submesh.indexCount,
                                                 indexType: submesh.indexType,
                                                 indexBuffer: indexBuffer.buffer,
                                                 indexBufferOffset: indexBuffer.offset)
        }
    }
    
    func GenCameraPosition() -> float3 {
        let distance: Float = 3.0
        let cosvalue: Float = cos(.pi / 12.0)
        let sinvalue: Float = sin(.pi / 12.0)
        var ret = Position
        ret += -distance * cosvalue * Cam.Front
        ret += distance * sinvalue * Cam.Up
        return ret
    }
    
    func CollisionCheck(scene: Scene) {
        collision = Collision(exist: false, Point: float3(), Normal: float3())
        var cx: Int = 0, cy: Int = 0, mx: Int = 0, my: Int = 0
        GetChunkMeshID(scene: scene, cx: &cx, cz: &cy, mx: &mx, mz: &my)
        dcx = cx
        dcy = cy
        dmx = mx
        dmy = my
        for i in -1 ... 2 {
            for j in -1 ... 2 {
                var meshx = mx + i
                var meshy = my + j
                var chunkx, chunky: Int
                if (meshx < 0) {
                    chunkx = cx - 1
                    meshx += MESH_SIZE
                } else if (meshx >= MESH_SIZE) {
                    chunkx = cx + 1
                    meshx -= MESH_SIZE
                } else {
                    chunkx = cx
                }
                
                if (meshy < 0) {
                    chunky = cy - 1
                    meshy += MESH_SIZE
                }
                else if (meshy >= MESH_SIZE) {
                    chunky = cy + 1
                    meshy -= MESH_SIZE
                }
                else {
                    chunky = cy
                }
                
                Maycol[i + 1][j + 1] = scene.chunk_offset[chunkx][chunky] + scene.mesh_offset[meshx][meshy]
                Maycol[i + 1][j + 1].y = scene.chunk[chunkx][chunky].height[meshx][meshy]
                if (Maycol[i + 1][j + 1].y < 0.1) {
                    Maycol[i + 1][j + 1].y = -0.5
                }
            }
        }

        var delta_pos = float3(0.0)
        for i in 0 ..< 3 {
            for j in 0 ..< 3 {
                var a, b, c: float3
                var dist: Float
                a = Maycol[i][j]
                b = Maycol[i + 1][j]
                c = Maycol[i + 1][j + 1]
                dist = distance(Position, a, b, c)
                if (dist <= Radius) {
                    collision.exist = true
                    collision.Normal += normalize(cross(c - a, b - a))
                    delta_pos += (Radius - dist) * collision.Normal
                }
                b = c
                c = Maycol[i][j + 1]
                dist = distance(Position, a, b, c)
                if (dist <= Radius) {
                    collision.exist = true
                    collision.Normal += normalize(cross(c - a, b - a))
                    delta_pos += (Radius - dist) * collision.Normal
                }
            }
        }
        Position += delta_pos
        if (collision.exist) {
            collision.Normal = normalize(collision.Normal)
        }
    }
    
    func GetChunkMeshID(scene: Scene, cx: inout Int, cz: inout Int, mx: inout Int, mz: inout Int) {
        var position = Position - scene.chunk[CHUNK_RADIUS][CHUNK_RADIUS].submesh[MESH_RADIUS][MESH_RADIUS].get_Position() - float3(MESH_LENGTH / 2.0, 0.0, MESH_LENGTH / 2.0)
        
        if (position.x > 0) {
            cx = Int(position.x / CHUNK_LENGTH + 0.5) + CHUNK_RADIUS
        } else {
            cx = Int(position.x / CHUNK_LENGTH - 0.5) + CHUNK_RADIUS
        }
        if (position.z > 0) {
            cz = Int(position.z / CHUNK_LENGTH + 0.5) + CHUNK_RADIUS
        } else {
            cz = Int(position.z / CHUNK_LENGTH - 0.5) + CHUNK_RADIUS
        }
        position.x -= Float(cx - CHUNK_RADIUS) * CHUNK_LENGTH
        position.z -= Float(cz - CHUNK_RADIUS) * CHUNK_LENGTH
        if (position.x > 0) {
            mx = Int(position.x / MESH_LENGTH + 0.5) + MESH_RADIUS
        } else {
            mx = Int(position.x / MESH_LENGTH - 0.5) + MESH_RADIUS
        }
        if (position.z > 0) {
            mz = Int(position.z / MESH_LENGTH + 0.5) + MESH_RADIUS
        } else {
            mz = Int(position.z / MESH_LENGTH - 0.5) + MESH_RADIUS
        }
    }
    
    func UpdatePosition(deltaTime: Float, scene: Scene) {
        Position += speed * deltaTime
        CollisionCheck(scene: scene)
        
        let angle = wspeed * deltaTime / 2.0
        let rot = simd_quatf(real: cos(angle),
                             imag: float3(sin(angle) * rotate_axis.x,
                                          sin(angle) * rotate_axis.y,
                                          sin(angle) * rotate_axis.z)
                            )
        rotate_state = rot * rotate_state
    }
    
    func UpdateSpeed(deltaTime: Float) {
        if (collision.exist) {
            let velocity = -dot(collision.Normal, speed)
            speed += velocity * collision.Normal
            if (velocity > 2.0) {
                speed += 0.1 * velocity * collision.Normal
            }
        }
        UpdateMovVec()
        
        var delta_v = float3(0.0)
        var noKeyboard = true
        if (collision.exist){
            if (ResourceManager.keys[BallMovement.forward.rawValue]) {
                let cosvalue = dot(Cam.Front, Mov.Up)
                delta_v += (Cam.Front - cosvalue * Mov.Up) * Acceleration.x * deltaTime
                noKeyboard = false
            }
            if (ResourceManager.keys[BallMovement.backward.rawValue]) {
                let cosvalue = dot(Cam.Front, Mov.Up)
                delta_v -= (Cam.Front - cosvalue * Mov.Up) * Acceleration.x * deltaTime
                noKeyboard = false
            }
            if (ResourceManager.keys[BallMovement.right.rawValue]) {
                let cosvalue = dot(Cam.Right, Mov.Up)
                delta_v += (Cam.Right - cosvalue * Mov.Up) * Acceleration.z * deltaTime
                noKeyboard = false
            }
            if (ResourceManager.keys[BallMovement.left.rawValue]) {
                let cosvalue = dot(Cam.Right, Mov.Up)
                delta_v -= (Cam.Right - cosvalue * Mov.Up) * Acceleration.z * deltaTime
                noKeyboard = false
            }
            if (ResourceManager.keys[BallMovement.jump.rawValue]) {
                delta_v += float3(0.0, 3.0, 0.0)
            }
            //Sliding Friction
            let movXZ = speed
            let movSpeed = length(movXZ)
            if (Double(movSpeed) > eps) {
                if (Resistance.x * deltaTime > movSpeed && noKeyboard) {
                    delta_v -= normalize(movXZ) * movSpeed
                } else {
                delta_v -= normalize(movXZ) * Resistance.x * deltaTime
                }
            }
        }
        //Gravity
        delta_v.y -= Resistance.y * deltaTime
        speed += delta_v
        
        UpdateMovVec()
        
        wspeed = length(speed) / Radius
        rotate_axis = -Mov.Right
    }
    
    func UpdateMovVec() {
        let speed = self.speed
        
        if(Double(abs(speed.x)) < eps && Double(abs(speed.x)) < eps) {
            Mov.Front = float3(1.0, 0.0, 0.0);
            Mov.Right = float3(0.0, 0.0, 1.0);
            Mov.Up    = float3(0.0, 1.0, 0.0);
        }
        else {
            Mov.Front = normalize(speed);
            Mov.Right = normalize(cross(Mov.Front, float3(0.0, 1.0, 0.0)));
            Mov.Up    = normalize(cross(Mov.Right, Mov.Front));
        }
    }
    
}
